File: android/app/src/main/java/com/github/gafiatulin/parakeetflow/core/util/FillerWordFilter.kt
package com.github.gafiatulin.parakeetflow.core.util

object FillerWordFilter {

    private val FILLER_PATTERN = Regex(
        """\b(uh|um|erm|ah|er|like|you know|I mean|sort of|kind of|basically|actually|literally|right|so|well|okay so)\b""",
        RegexOption.IGNORE_CASE
    )

    private val REPEATED_WORD_PATTERN = Regex(
        """\b(\w+)\s+\1\b""",
        RegexOption.IGNORE_CASE
    )

    fun filter(text: String): String {
        var result = text
        result = FILLER_PATTERN.replace(result, "")
        result = REPEATED_WORD_PATTERN.replace(result, "$1")
        result = result.replace(Regex("""\s{2,}"""), " ").trim()
        result = result.replace(Regex("""\s+([.,!?;:])"""), "$1")
        return result
    }
}

File: mac/Sources/ParakeetFlow/ASR/FillerWordFilter.swift
import Foundation

enum FillerWordFilter {
    static let defaultFillerWords: [String] = [
        "you know",
        "I mean",
        "kind of",
        "sort of",
        "um",
        "uh",
        "ah",
        "er",
        "like",
        "basically",
        "actually",
        "literally",
        "honestly",
    ]

    nonisolated(unsafe) private static var compiledPatterns: [(NSRegularExpression, String)] = {
        buildPatterns(defaultFillerWords)
    }()

    static func updatePatterns(_ words: [String]) {
        compiledPatterns = buildPatterns(words)
    }

    private static func buildPatterns(_ words: [String]) -> [(NSRegularExpression, String)] {
        words.compactMap { word in
            let pattern = "\\b\(NSRegularExpression.escapedPattern(for: word))\\b"
            guard let regex = try? NSRegularExpression(pattern: pattern, options: .caseInsensitive) else {
                return nil
            }
            return (regex, word)
        }
    }

    static func removeFillersFromText(_ text: String) -> String {
        var result = text
        for (regex, _) in compiledPatterns {
            result = regex.stringByReplacingMatches(
                in: result,
                range: NSRange(result.startIndex..., in: result),
                withTemplate: ""
            )
        }
        // Collapse multiple spaces
        while result.contains("  ") {
            result = result.replacingOccurrences(of: "  ", with: " ")
        }
        return result.trimmingCharacters(in: .whitespacesAndNewlines)
    }
}
